### PR TITLE
Add typechain to nitro-protocol

### PR DESCRIPTION
**Pull request by geoknee**
https://www.notion.so/statechannels/Add-TypeChain-tooling-f17342226cb44244be4820b9937c9fae

This PR brings us to a state where we can do this: 

Screenshot 2021-06-07 at 13 11 49

I didn't spread the types around the codebase much, yet. For now I just used them in the gas benchmarking test. We can do more of that kind of thing in future PRs.

### CODE REVIEW

File: packages/nitro-protocol/gas-benchmarks/vanillaSetup.ts [hunk 1]
```diff
@@ -21,10 +24,10 @@ declare global {
   }
 }
 
-export let ethAssetHolder: Contract;
-export let erc20AssetHolder: Contract;
-export let nitroAdjudicator: Contract;
-export let token: Contract;
+export let ethAssetHolder: ETHAssetHolder & Contract;
```
Comment (geoknee) [inline packages/nitro-protocol/gas-benchmarks/vanillaSetup.ts:27]: I don't think this intersection type is the canonical way to use typechain. But it works well!

File: .circleci/config.yml [hunk 2]
```diff
@@ -112,6 +112,7 @@ jobs:
             - packages/*/dist
             - packages/*/build
             - packages/nitro-protocol/artifacts
+            - packages/nitro-protocol/typechain
```
Comment (NiloCK) [inline .circleci/config.yml:115]: Clarifying circle config:

This is writing these directories into the filesystem of circleCI's vm, and not persisting in the 'artifact' sense? It writes these to make them available for the later test job.

Comment (geoknee) [reply .circleci/config.yml:115]: Exactly!

**Review by NiloCK — APPROVED** [review]
This is great! The gas benchmarks, along with typechain, serve as a much improved walk-through of nitro workflows with great discoverability.

I ~am having~ had some trouble running this locally. For reference, things were cleared up after some `yarn clean`, `yarn`, and closing out *all* VSCode windows, which seemed to be hanging onto some out of date cached typings.

**Pull request merged**